### PR TITLE
fix: add more tests for tiering extension [1]

### PR DESCRIPTION
**Pull request by kasrakhosravi**
## Description:

> Adding more test cases for tiering

## PR Checklist
- [ ] :eyes: I have self-reviewed my PR, leaving comments on any changes which may trigger a discussion
- [ ] :pencil2: Added relevant tests

### CODE REVIEW

File: contracts/collections/ERC721/extensions/ERC721TieringExtension.sol [hunk 1]
```diff
@@ -214,6 +216,10 @@ abstract contract ERC721TieringExtension is
         view
         returns (uint256 tierRemaining)
     {
+        // remaining supply cannot be less than remaining reserved spots
+         if (maxSupply - totalSupply() < totalReserved - reservedMints) {
+            revert OverflowSupply(maxSupply - totalSupply(), totalReserved - reservedMints);
+        }
```
Comment (aramalipoor) [inline contracts/collections/ERC721/extensions/ERC721TieringExtension.sol:222]: I think we should put checks in other places so we don't end up in such situation. Here is too late to check for this 🤔

File: test/collections/presets/ERC721TieredSalesCollection.test.ts [hunk 2]
```diff
@@ -1235,22 +1235,6 @@ describe("ERC721TieredSalesCollection", function () {
...
-          value: utils.parseEther("0.24"),
-        });
-
-        await expect(
-          collection.connect(userA.signer).mintByTier(0, 2, 0, [], {
-            value: utils.parseEther("0.12"),
-          })
-        ).to.be.revertedWith("EXCEEDS_MAX");
-      });
-
```
Comment (aramalipoor) [inline test/collections/presets/ERC721TieredSalesCollection.test.ts:1253]: Don't see if this is covered in other new tests. Basically edge-cases around per-wallet allocation configured globally on a tier. Is it already covered?

File: contracts/collections/ERC721/extensions/ERC721TieringExtension.sol [hunk 3]
```diff
@@ -42,6 +42,8 @@ interface IERC721TieringExtension {
         uint256 maxAllowance,
         bytes32[] calldata proof
     ) external payable;
+
+    error OverflowSupply(uint256 supply, uint256 reserve);
```
Comment (aramalipoor) [inline contracts/collections/ERC721/extensions/ERC721TieringExtension.sol:46]: As we discussed better to adopt errors in a different PR

File: test/collections/presets/ERC721TieredSalesCollection.test.ts (hunk 2 above)
Comment (kasrakhosravi) [reply test/collections/presets/ERC721TieredSalesCollection.test.ts:1257]: @aramalipoor I think it is the exact same test of `should fail when minting a tier and amount + prev mints is asking for more than remaining per-wallet allocation`. can you check this to make sure I deleted it correctly?

File: contracts/collections/ERC721/extensions/ERC721TieringExtension.sol (hunk 1 above)
Comment (kasrakhosravi) [reply contracts/collections/ERC721/extensions/ERC721TieringExtension.sol]: yeah this is the check we had in the `setMaxSupply` in the other PR. now this is removed

File: test/collections/presets/ERC721TieredSalesCollection.test.ts (hunk 2 above)
Comment (aramalipoor) [reply test/collections/presets/ERC721TieredSalesCollection.test.ts:1257]: Oh yeah that's correct, it's duplicate

File: test/collections/presets/ERC721TieredSalesCollection.test.ts [hunk 4]
```diff
@@ -1681,6 +1665,309 @@ describe("ERC721TieredSalesCollection", function () {
...
+            merkleRoot: mkt2.getHexRoot(),
+            reserved: 2,
+            maxAllocation: 5000,
+          });
+
+        await collection.connect(userA.signer).mintByTier(
+          1,
+          1,
+          1,
+          mkt1.getHexProof(
```
Comment (aramalipoor) [inline test/collections/presets/ERC721TieredSalesCollection.test.ts:1870]:
```suggestion
          mkt2.getHexProof(
```
?

**Review by aramalipoor — APPROVED** [review]
